flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java

```diff
@@ -0,0 +1,134 @@
+  private final KeySelector<T, K> keySelector;
+  private final OperatorEventGateway operatorEventGateway;
+  private final DataStatisticsFactory<K> statisticsFactory;
+  private transient volatile DataStatistics<K> localStatistics;
+  private transient volatile DataStatistics<K> globalStatistics;
+  private transient ListState<DataStatistics<K>> globalStatisticsState;
+
+  DataStatisticsOperator(
+      KeySelector<T, K> keySelector,
+      OperatorEventGateway operatorEventGateway,
+      DataStatisticsFactory<K> statisticsFactory) {
+    this.keySelector = keySelector;
+    this.operatorEventGateway = operatorEventGateway;
+    this.statisticsFactory = statisticsFactory;
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    localStatistics = statisticsFactory.createDataStatistics();
+    globalStatisticsState =
+        context
+            .getOperatorStateStore()
+            .getUnionListState(
+                new ListStateDescriptor<>(
+                    "globalStatisticsState",
+                    TypeInformation.of(new TypeHint<DataStatistics<K>>() {})));
+
+    if (context.isRestored()) {
+      int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+      if (globalStatisticsState.get() == null
+          || !globalStatisticsState.get().iterator().hasNext()) {
+        LOG.warn("Subtask {} doesn't have global statistics state to restore", subtaskIndex);
+        globalStatistics = statisticsFactory.createDataStatistics();
+      } else {
+        LOG.info("Restoring global statistics state for subtask {}", subtaskIndex);
+        globalStatistics = globalStatisticsState.get().iterator().next();
+      }
+    } else {
+      globalStatistics = statisticsFactory.createDataStatistics();
+    }
+  }
+
+  @Override
+  public void open() throws Exception {
+    if (!globalStatistics.isEmpty()) {
+      output.collect(new StreamRecord<>(DataStatisticsOrRecord.fromDataStatistics(globalStatistics)));
+    }
+  }
+
+  @Override
+  public void handleOperatorEvent(OperatorEvent evt) {
+    // TODO: receive event with aggregated statistics from coordinator and update globalStatistics
+  }
+
+  @Override
+  public void processElement(StreamRecord<T> streamRecord) throws Exception {
+    final K key = keySelector.getKey(streamRecord.getValue());
+    localStatistics.add(key);
+    output.collect(new StreamRecord<>(DataStatisticsOrRecord.fromRecord(streamRecord.getValue())));
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    long checkpointId = context.getCheckpointId();
+    LOG.info("Taking data statistics operator snapshot for checkpoint {}", checkpointId);
+
+    // Only subtask 0 saves the state so that globalStatisticsState(UnionListState) stores
+    // an exact copy of globalStatistics
+    if (!globalStatistics.isEmpty() && getRuntimeContext().getIndexOfThisSubtask() == 0) {
```

Comment:
it will change after first checkpoint. globalStatistics is being updated in the function #handleOperatorEvent when operator receives the event from coordinator. 
The purpose of adding `!globalStatistics.isEmpty` is, in the first round of checkpoint, operator doesn't have any global statistics yet, when taking snapshot, it will store the empty `globalStatistics`  into `globalStatisticsState`

Reply:
you are right. I misread the code.